ODLA/platforms/dnnl/odla_dnnl.cc

```diff
@@ -1508,6 +1513,28 @@ static odla_value reduce_op(dnnl::algorithm alg, odla_value input,
   return v;
 }
 
+odla_value odla_ReduceLogSum(odla_value input, odla_size_t num_of_axes,
+                             const odla_uint32* axes, odla_bool keep_dims,
+                             odla_value_shape output_dims,
+                             const odla_value_id id) {
+  const auto& name = std::string(reinterpret_cast<const char*>(id)) + "_extra";
+  return odla_Log(
+      odla_ReduceSum(input, num_of_axes, axes, keep_dims, output_dims,
+                     (const odla_value_id)name.c_str()),
+      id);
+}
+
+odla_value odla_ReduceLogSumExp(odla_value input, odla_size_t num_of_axes,
```

Comment:
do we need to consider overflow and underflow issue? 
https://www.xarg.org/2016/06/the-log-sum-exp-trick-in-machine-learning/

Reply:
let me give a try.

Reply:
I tried the following code:
```
 auto reduce_max = odla_ReduceMax(input, num_of_axes, axes, keep_dims,
                                   output_dims, (const odla_value_id) "name1");
  auto reduce_max_keep_dim =
      odla_Reshape(reduce_max, keep_dim_shape, (const odla_value_id) "name2");
  auto exp_delta = odla_Exp(
      odla_Sub(input, reduce_max_keep_dim, (const odla_value_id) "name4"),
      (const odla_value_id) "name3");
  return odla_Add(odla_ReduceLogSum(exp_delta, num_of_axes, axes, keep_dims,
                                    output_dims, (const odla_value_id) "name5"),
                  reduce_max, id);
```

The problem is: if the reduction axes happens at left, for example { 3,4 ,5 } => {1, 1, 5}, the binary op will fail.
This is because DNNL only support broadcasting from right.

Reply:
ok, i c. lets stick to the original solution.